src/Forms/FileUploadable.php

```diff
@@ -0,0 +1,399 @@
+ *   - many_many
+ * Additionally supports writing directly to the File table not attached
+ * to any parent record.
+ *
+ * @mixin FormField
+ */
+trait FileUploadable {
+	use Uploadable;
+
+	/**
+	 * Flag to automatically determine and save a has_one-relationship
+	 * on the saved record (e.g. a "Player" has_one "PlayerImage" would
+	 * trigger saving the ID of newly created file into "PlayerImageID"
+	 * on the record).
+	 *
+	 * @var boolean
+	 */
+	public $relationAutoSetting = true;
+
+	/**
+	 * Parent data record. Will be infered from parent form or controller if blank.
+	 *
+	 * @var DataObject
+	 */
+	protected $record;
+
+	/**
+	 * Items loaded into this field. May be a RelationList, or any other SS_List
+	 *
+	 * @var SS_List
+	 */
+	protected $items;
+
+	protected function constructFileUploadable() {
+		$this->constructUploadable();
+	}
+
+
+	/**
+	 * Force a record to be used as "Parent" for uploaded Files (eg a Page with a has_one to File)
+	 *
+	 * @param DataObject $record
+	 * @return $this
+	 */
+	public function setRecord($record) {
+		$this->record = $record;
+		return $this;
+	}
+	/**
+	 * Get the record to use as "Parent" for uploaded Files (eg a Page with a has_one to File) If none is set, it will
+	 * use Form->getRecord() or Form->Controller()->data()
+	 *
+	 * @return DataObject
+	 */
+	public function getRecord() {
+		if ($this->record) {
+			return $this->record;
+		}
+		if (!$this->getForm()) {
+			return null;
+		}
+
+		// Get record from form
+		$record = $this->getForm()->getRecord();
+		if ($record && ($record instanceof DataObject)) {
+			$this->record = $record;
+			return $record;
+		}
+
+		// Get record from controller
+		$controller = $this->getForm()->getController();
+		if ($controller
+			&& $controller->hasMethod('data')
+			&& ($record = $controller->data())
+			&& ($record instanceof DataObject)
+		) {
+			$this->record = $record;
+			return $record;
+		}
+
+		return null;
+	}
+
+
+	/**
+	 * Loads the related record values into this field. This can be uploaded
+	 * in one of three ways:
+	 *
+	 *  - By passing in a list of file IDs in the $value parameter (an array with a single
+	 *    key 'Files', with the value being the actual array of IDs).
+	 *  - By passing in an explicit list of File objects in the $record parameter, and
+	 *    leaving $value blank.
+	 *  - By passing in a dataobject in the $record parameter, from which file objects
+	 *    will be extracting using the field name as the relation field.
+	 *
+	 * Each of these methods will update both the items (list of File objects) and the
+	 * field value (list of file ID values).
+	 *
+	 * @param array $value Array of submitted form data, if submitting from a form
+	 * @param array|DataObject|SS_List $record Full source record, either as a DataObject,
+	 * SS_List of items, or an array of submitted form data
+	 * @return $this Self reference
+	 * @throws ValidationException
+	 */
+	public function setValue($value, $record = null) {
```

Comment:
This will silently fail if applied to a `FormField` which already implements `setValue()`, due to how [trait precedence](https://www.sitepoint.com/using-traits-in-php-5-4/) works. In the case of `AssetField-> setValue()`, it seems to be overruled on purpose - which makes this even more confusing. Same with `saveInto()`. Shouldn't this live on `UploadField` if its only used there? I'm starting to think the trait does more harm than good in this case: It avoid a bit of trivial duplication at the expense of a much less readable implementation.